pkg/services/control/server/server.go

```diff
@@ -65,89 +72,34 @@ type cfg struct {
 
 	treeService TreeService
 
-	s *engine.StorageEngine
-}
-
-func defaultCfg() *cfg {
-	return &cfg{}
+	storage *engine.StorageEngine
 }
 
 // New creates, initializes and returns new Server instance.
-func New(opts ...Option) *Server {
-	c := defaultCfg()
-
-	for _, opt := range opts {
-		opt(c)
+// Must be marked as available with [Server.MarkReady] when all the
+// components for serving are ready. Before [Server.MarkReady] call
+// only health checks are available.
+func New(key *ecdsa.PrivateKey, authorizedKeys [][]byte, healthChecker HealthChecker) *Server {
+	cfg := &cfg{
+		key:           key,
+		allowedKeys:   authorizedKeys,
+		healthChecker: healthChecker,
 	}
 
 	return &Server{
-		cfg: c,
+		cfg: cfg,
 	}
 }
 
-// WithKey returns option to set private key
-// used for signing responses.
-func WithKey(key *ecdsa.PrivateKey) Option {
-	return func(c *cfg) {
-		c.key = key
-	}
-}
-
-// WithAuthorizedKeys returns option to add list of public
-// keys that have rights to use Control service.
-func WithAuthorizedKeys(keys [][]byte) Option {
-	return func(c *cfg) {
-		c.allowedKeys = append(c.allowedKeys, keys...)
-	}
-}
-
-// WithHealthChecker returns option to set component
-// to calculate node health status.
-func WithHealthChecker(hc HealthChecker) Option {
-	return func(c *cfg) {
-		c.healthChecker = hc
-	}
-}
-
-// WithNetMapSource returns option to set network map storage.
-func WithNetMapSource(netMapSrc netmap.Source) Option {
-	return func(c *cfg) {
-		c.netMapSrc = netMapSrc
-	}
-}
-
-// WithContainerSource returns option to set container storage.
-func WithContainerSource(cnrSrc container.Source) Option {
-	return func(c *cfg) {
-		c.cnrSrc = cnrSrc
-	}
-}
-
-// WithReplicator returns option to set network map storage.
-func WithReplicator(r *replicator.Replicator) Option {
-	return func(c *cfg) {
-		c.replicator = r
-	}
-}
-
-// WithNodeState returns option to set node network state component.
-func WithNodeState(state NodeState) Option {
-	return func(c *cfg) {
-		c.nodeState = state
-	}
-}
-
-// WithLocalStorage returns option to set local storage engine that
-// contains information about shards.
-func WithLocalStorage(engine *engine.StorageEngine) Option {
-	return func(c *cfg) {
-		c.s = engine
-	}
-}
-
-// WithTreeService returns an option to set tree service.
-func WithTreeService(s TreeService) Option {
-	return func(c *cfg) {
-		c.treeService = s
-	}
+// MarkReady marks server available. Before this call none of the other calls
+// are available except for the health checks.
+func (s *Server) MarkReady(e *engine.StorageEngine, nm netmap.Source, c container.Source, r *replicator.Replicator, st NodeState, tr TreeService) {
```

Comment:
suggest to require all prms to be non-nil and insta panic otherwise

Reply:
fixed